utils/FileHandler.cpp

```diff
@@ -0,0 +1,315 @@
+        QString firstLine;  // Stores the first line of the file
+
+        if(!file.open(QIODevice::ReadOnly))
+        {
+           // If the file didn't open, throw an exception
+           throw BadFile(file);
+        }
+
+        // Expect an empty line at the beginning of the file
+        firstLine = qin.readLine();
+        if(!firstLine.isEmpty())
+        {
+            throw BadFileFormat(file, "empty line", firstLine);
+        }
+
+        // Push each shape extracted back into the vector until we get to the end of the file
+        while(!qin.atEnd())
+        {
+            // Extract shape throws if the shape could not be identified
+            shapeExtracted = extractShape(qin);
+            shapesToReturn.push_back(shapeExtracted);
+        }
+
+        // Close the file when loading is finised
+        file.close();
+    }// end try
+    catch (GeneralException& exception) {
+        exception.errorWindow();
+    }
+
+    return shapesToReturn;
+}
+
+Shape* FileHandler::extractShape(QTextStream& qin)
+{
+    QString shapeType;  // Name of the shape being extracted currently
+    QList<int> dimensionsList;   // List of dimensions of the shape
+    Shape* shape;    // Pointer to the shape that is returned by the function
+
+    // Ignore the shape's ID data
+    qin.readLine();
+
+    // Get the name of the shape to add
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    qin >> shapeType;
+
+    // Throw an exception if the shape type extracted is not in the list of valid shape types
+    if(jconstants::TYPE_LIST.indexOf(shapeType) == -1)
+    {
+        QFile* file = dynamic_cast<QFile*>(qin.device());
+
+        if(file != nullptr)
+        {
+            throw BadFileFormat(*file, "a valid shape type name", shapeType);
+        }
+        else
+        {
+            throw BadFileFormat("a valid shape type name", shapeType);
+        }
+    }
+
+    // Extract the dimensions from the file
+    dimensionsList = extractDimensions(qin, shapeType);
+
+    // If the shape being read in isn't text...
+    if(shapeType != jconstants::TEXT)
+    {
+        //...call corresponding extraction function
+        shape = extractNonText(qin, shapeType, dimensionsList);
+    }
+    // Otherwise, if the shape is a text...
+    else
+    {
+        //...call the corresponding extraction function
+        shape = extractText(qin, dimensionsList);
+    }
+
+    // Read two more lines to get to the next shape's data
+    qin.readLine();
+    qin.readLine();
+
+    return shape;
+}
+
+Shape* FileHandler::extractNonText(QTextStream& qin, QString& shapeType, QList<int>& dimensionsList)
+{
+    Shape* shape;    // Shape pointer to be returned by the function
+
+    // Extract data for the pen that draws this shape' edges
+    extractPen(qin);
+
+    // If the shape being read in isn't a line, we know it needs a brush
+    if(shapeType != jconstants::LINE && shapeType != jconstants::POLYLINE)
```

Comment:
Can we reverse this ``if`` condition?
```suggestion
    if(shapeType == jconstants::LINE || shapeType == jconstants::POLYLINE)
    { /* Do stuff for line and polyline */ }
```

Reply:
I guess? Seems like a nitpick

Reply:
But I'll do it anyway if it gets you to FREACKING APPROVE THE PULL 😉